[llvm] Support multiple save/restore points in mir

enoskova-sc opened the pull request:
Currently mir supports only one save and one restore point specification:

```
  savePoint:       '%bb.1'
  restorePoint:    '%bb.2'
```

This patch provide possibility to have multiple save and multiple restore points in mir:

```
  savePoints:
    - point:           '%bb.1'
  restorePoints:
    - point:           '%bb.2'
```

Shrink-Wrap points split Part 3.
RFC: https://discourse.llvm.org/t/shrink-wrap-save-restore-points-splitting/83581

Part 1: https://github.com/llvm/llvm-project/pull/117862
Part 2: https://github.com/llvm/llvm-project/pull/119355
Part 4: https://github.com/llvm/llvm-project/pull/119358
Part 5: https://github.com/llvm/llvm-project/pull/119359

enoskova-sc commented:
@arsenm, could you take a look, please.

preames commented on llvm/test/CodeGen/AArch64/GlobalISel/store-merging-debug.mir:
```diff
@@ -86,8 +86,8 @@ frameInfo:
   localFrameSize:  0
-  savePoint:       ''
-  restorePoint:    ''
+  savePoints:      []
```
You will cut out a huge amount of spurious diff if you teach the parser to support both the singular and plural forms here.

enoskova-sc replied:
Аddressed. Now MIR can be consumed both in singular:
` savePoint:    '%bb.1'`
 and plural:
```
savePoint:
    - point:           '%bb.1'
```
   formats.
But printed in only plural format.

arsenm commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -618,6 +616,18 @@ void MIRPrinter::convert(yaml::MachineFunction &MF,
+                         MachineBasicBlock *SRP) {
+  std::string Str;
+  yaml::SRPEntry Entry;
+  raw_string_ostream StrOS(Str);
```
raw_svector_ostream + SmallVector

arsenm commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -618,6 +616,18 @@ void MIRPrinter::convert(yaml::MachineFunction &MF,
+  raw_string_ostream StrOS(Str);
+  StrOS << printMBBReference(*SRP);
+  Entry.Point = StrOS.str();
+  Str.clear();
```
Don't need this

enoskova-sc replied:
addressed

michaelmaitland commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -610,6 +610,20 @@ LLVM_YAML_IS_SEQUENCE_VECTOR(llvm::yaml::MachineJumpTable::Entry)
 namespace llvm {
 namespace yaml {
 
+struct SRPEntry {
```
nit: rename to `SaveRestorePointEntry`? I don't think SRP is a common acronym. Also please add a docstring.

enoskova-sc replied:
addressed

michaelmaitland commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -832,18 +836,9 @@ bool MIRParserImpl::initializeFrameInfo(PerFunctionMIParsingState &PFS,
-    MFI.setRestorePoint(MBB);
-  }
+  initializeSaveRestorePoints(PFS, YamlMFI.SavePoints, true /*IsSavePoints*/);
+  initializeSaveRestorePoints(PFS, YamlMFI.RestorePoints,
```
We used to return true when `parseMBBReference`. Are we missing this functionality now?

enoskova-sc replied:
Addressed, now this functionality isn't missed.

michaelmaitland commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1058,8 +1053,28 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
 
-bool MIRParserImpl::initializeJumpTableInfo(PerFunctionMIParsingState &PFS,
-    const yaml::MachineJumpTable &YamlJTI) {
+bool MIRParserImpl::initializeSaveRestorePoints(
```
Add docstring clarifying what it means when true/false is returned?

enoskova-sc replied:
addressed

michaelmaitland commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1058,8 +1053,28 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+  return false;
+}
+
+bool MIRParserImpl::initializeJumpTableInfo(
```
Looks like formatting unrelated to this patch occurred?

enoskova-sc replied:
addressed

michaelmaitland commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -392,14 +394,10 @@ void MIRPrinter::convert(ModuleSlotTracker &MST,
-    raw_string_ostream StrOS(YamlMFI.RestorePoint.Value);
-    StrOS << printMBBReference(*MFI.getRestorePoint());
-  }
+  if (MFI.getSavePoint())
```
Here we have singular getSavePoint, plural YamlMFI.SavePoints, and singular getSavePoint again. It is a little weird that we're mixing singular and plural.

This makes me think that we should get rid of SavePoint and RestorePoint and replace it entirely with the plural SavePoints and RestorePoints. By this I mean `SavePoint` and `RestorePoint` should be replaced entirely in MIR with `SavePoints` and `RestorePoints`. There plural version can do everything the singular version can and more.

This could be done in a follow up patch if others agree this would be a good direction to move in.

michaelmaitland commented:
Reverse ping!

enoskova-sc commented:
@michaelmaitland, I was on long vacations. This week I plan to return to work on this MR:)

michaelmaitland commented:
> @michaelmaitland, I was on long vacations. This week I plan to return to work on this MR:)

Lovely, I hope you had a wonderful vacation! Excited to give review and drive this forward.

enoskova-sc committed:
[llvm] support multiple save/restore points in mir

Currently mir supports only one save and one restore point specification:

```
  savePoint:       '%bb.1'
  restorePoint:    '%bb.2'
```

This patch provide possibility to specify multiple save and multiple restore points in mir:

```
  savePoint:
    - point:           '%bb.1'
  restorePoint:
    - point:           '%bb.2'
```
while maintaining backward compatibility.

michaelmaitland reviewed (commented):
Should we have a test that shows multiple save points and multiple restore points?

enoskova-sc commented:
@michaelmaitland, @preames, @arsenm, I have addressed everything. Do you have any other comments?

preames commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1077,6 +1075,41 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+    if (VectorRepr.empty())
+      return false;
+
+    const auto &Entry = VectorRepr.front();
```
Unless I'm missing something from the code structure, you're only parsing the first bb reference here.  Also, you don't seem to have any tests for round tripping MIR with multiple save restore points.

enoskova-sc replied:
Test was added, code fixed

michaelmaitland commented:
> @michaelmaitland, now multiple save and restore points are not supported in shrink-wrap. This patch only support possibility to print them in MIR. Real multiple save/restore points can appear only in #119359.

Can we have parser tests to show that it can consume multiple save/restore points?

michaelmaitland commented on llvm/test/tools/llvm-reduce/mir/preserve-frame-info.mir:
```diff
@@ -117,7 +119,7 @@ frameInfo:
   localFrameSize:  0
   savePoint:       '%bb.1'
-  restorePoint:    '%bb.2'
+  restorePoint:    '%bb.1'
```
Is this change related?

enoskova-sc replied:
removed this change

michaelmaitland commented on llvm/test/CodeGen/MIR/X86/frame-info-multiple-save-restore-points-parse.mir:
```diff
@@ -0,0 +1,81 @@
+
+--- |
+
+  define i32 @foo(i32 %a, i32 %b) {
```
Is the IR section needed?

enoskova-sc commented:
In my new commit I added support for "multiple" save/restore points in MFI, because without it I can't create a test.
The added test is quite synthetic, but it checks the capabilities of the MIRParser and MIRPrinter.

michaelmaitland commented on llvm/test/CodeGen/MIR/X86/frame-info-multiple-save-restore-points-parse.mir:
```diff
@@ -0,0 +1,81 @@
+  maxAlignment:  4
+  hasCalls:      true
+  savePoint:
+    - point:     '%bb.1'
```
Thanks for pointing me to that follow up patch. I think I would prefer to see this PR contain information about the registers despite this PR becoming more complicated. My reasoning is based on two things:

1. It is difficult to write a test for just this PR without it.
2. The parsing and printing here is missing necessary information (registers) that is needed for multiple save/restore points. What does it mean to have MIR with multiple save/restore points without registers associated (garbage?)?
3. It looks like a large portion of diff in #119358 is related to test cases, not actual code.

@preames do you have an opinion here?

michaelmaitland replied:
I think we should combine the two PRs.

> The parsing and printing here is missing necessary information (registers) that is needed for multiple save/restore points. What does it mean to have MIR with multiple save/restore points without registers associated (garbage?)?

This point is the main reason why. This PR, without specifying which registers are saved at which restore points create an ambiguity for shrink wrapping. Currently shrink wrapping only supports save/restore to a single basic block. What does it mean for shrink wrapping when there is multiple save/restore blocks (i.e. this PR is merged)?

michaelmaitland replied:
Does this test need updating? What do the points correspond to? registers? frame idxs?

enoskova-sc replied:
test is updated, now it contains registers

enoskova-sc commented:
@michaelmaitland, do you want me to update this MR, adding information about registers, or should we wait for @preames to leave an opinion?

michaelmaitland commented:
> @michaelmaitland, do you want me to update this MR, adding information about registers, or should we wait for @preames to leave an opinion?

I thought I responded to this. Sorry. I'd like to see it updated in this PR to avoid the listed concerns.

enoskova-sc commented:
@michaelmaitland, I'll do it soon.

michaelmaitland commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -295,6 +322,10 @@ class MachineFrameInfo {
   /// Has CSInfo been set yet?
   bool CSIValid = false;
 
+  CalleeSavedInfoPerBB CSInfoPerSave;
```
If we combine SavePoints and CalleeSavedInfoPerBB and also combine RestorePoints with CSInfoPerRestore, then we may be able to drop these objects and save memory?

enoskova-sc replied:
Adressed. Now we have one structure called SaveRestorePoints.

michaelmaitland commented:
@enoskova-sc Thanks for combining the two PR's this is looking really great!

michaelmaitland commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1077,6 +1075,54 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+      return false;
+    if (parseMBBReference(PFS, MBB, StringRepr))
+      return true;
+    SRPoints.insert(std::make_pair(MBB, Registers));
```
Registers will be empty here. Will this mean that **all CSRs** must be spilled/restored or **no CSRs** must be spilled?  It probably means that **all CSRs** must be spilled/restored in this MBB. Maybe this should be documented on the data structure definition?

enoskova-sc replied:
addressed

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -478,12 +478,59 @@ bool RISCVRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
+        return CS.getFrameIdx() == FrameIndex;
+      });
+
+      if (It != CSI.end()) {
```
IS it possible for this condition to be false? If so, will SpilledIn and RestoredIn be nullptr below? I think that would be an issue. If it cannot be false, maybe we should drop the `if` statement here.

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -478,12 +478,59 @@ bool RISCVRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
+          SpilledIn = *It->spilledIn().begin();
+
+        else if (MI.mayLoad() && !It->restoredIn().empty())
+          RestoredIn = *It->restoredIn().begin();
```
Should this be `restoredIn().end()` in order to get the last basic block it needs to be restored in?

enoskova-sc replied:
no longer actual

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:
```diff
@@ -922,11 +966,10 @@ void RISCVFrameLowering::emitPrologue(MachineFunction &MF,
-  // directives.
-  emitCFIForCSI<CFISaveRegisterEmitter>(MBB, MBBI, getUnmanagedCSI(MF, CSI));
+  int Distance = getUnmanagedCSI(MF, CSI).size();
+  if (!RVFI->isPushable(MF) && !RVFI->useSaveRestoreLibCalls(MF))
```
Can you help me understand why this is needed for this patch?

enoskova-sc replied:
This piece of code originally aimed to skip stores of CSRs and right after these stores  emit one CFI instruction for one store.
As in this patch I moved CFI instruction emission to the `spillCalleeSavedRegisters` (because spills can appear not only in Prolog) here we need to remove CFI emission and skip 2*(num of CSRs): 1 for spill + 1 for CFI.

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:
```diff
@@ -1087,13 +1130,16 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
   auto FirstScalarCSRRestoreInsn =
       std::next(MBBI, getRVVCalleeSavedInfo(MF, CSI).size());
+  int Distance = getUnmanagedCSI(MF, CSI).size();
+  auto LastFrameDestroy = std::prev(MBBI, Distance);
```
Can you help me understand why this is needed for this patch?

enoskova-sc replied:
no longer actual

enoskova-sc commented:
@michaelmaitland, tank you for quick review and sorry for delay, I have a lot of another important tasks. I'm still working on this patch and will publish changes, as soon as possible.

arsenm commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -631,6 +631,61 @@ LLVM_YAML_IS_SEQUENCE_VECTOR(llvm::yaml::CalledGlobal)
+  }
+};
+
+using SaveRestorePoints =
+    std::variant<std::vector<SaveRestorePointEntry>, StringValue>;
```
I don't understand why this is a variant, and not just the vector of points

enoskova-sc replied:
Several comments above (https://github.com/llvm/llvm-project/pull/119357#discussion_r1880985027) @preames suggested to make support for multiple save/restore points backward compatible with single save/restore point approach.
It helped to "cut out a huge amount of spurious diff". So, StringValue in this variant needed for backward compatibility.

arsenm replied:
We should make no effort to maintain MIR backwards compatibility. As a diff reduction strategy, this can be part 1. But we should absolutely rip out the old form

preames replied:
I know I suggested the variant, but looking at the code, the complexity added seems like a bad tradeoff.

I'd be tempted to do a small change which "manually" parses and prints a single entry array, and updates all the tests to the new format.  If you did that, you could rebase this one on top and add the non-single entry general case.

I'll note that I'm explicitly okay with leaving this as is.  We can land this patch, then remove the complexity and do the test change in a follow up.  (Note to other reviews, I'm committing to write that patch if needed.)

arsenm commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -720,10 +784,14 @@ template <> struct MappingTraits<MachineFrameInfo> {
+    YamlIO.mapOptional(
+        "savePoint", MFI.SavePoints,
+        SaveRestorePoints(
+            StringValue())); // Don't print it out when it's empty.
```
Empty list?

arsenm commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -74,6 +74,34 @@ class CalleeSavedInfo {
+  void set(PointsMap CSI) { Map = std::move(CSI); }
+
+  MachineBasicBlock *findAny(const CalleeSavedInfo &Match) const {
+    for (auto [BB, CSIV] : Map)
```
braces

enoskova-sc replied:
addressed

arsenm commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -74,6 +74,34 @@ class CalleeSavedInfo {
+public:
+  const PointsMap &get() const { return Map; }
+
+  const std::vector<CalleeSavedInfo> getCSInfo(MachineBasicBlock *MBB) const {
```
Avoid vector copy, just return ArrayRef?

enoskova-sc replied:
Lookup here returns a default constructed vector, if MBB isn't found, ref on this vector is dangling.

arsenm replied:
Replace with at? there should be no default construction

enoskova-sc replied:
`at` here is not a good solution, because sometimes MBB really not present in map and it is not an error.
For example, not every Prolog is a Save BB (we don't save registers in it), but in emitPrologue we call getCSInfo() for Prolog.

arsenm commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -74,6 +74,34 @@ class CalleeSavedInfo {
+    return Map.lookup(MBB);
+  }
+
+  void set(PointsMap CSI) { Map = std::move(CSI); }
```
missing &&??

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -851,18 +855,12 @@ bool MIRParserImpl::initializeFrameInfo(PerFunctionMIParsingState &PFS,
-    MFI.setRestorePoint(MBB);
-  }
+  if (initializeSaveRestorePoints(PFS, YamlMFI.SavePoints,
+                                  true /*IsSavePoints*/))
```
```suggestion
                                  /*IsSavePoints=*/true))
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/MachineFrameInfo.cpp:
```diff
@@ -244,6 +244,23 @@ void MachineFrameInfo::print(const MachineFunction &MF, raw_ostream &OS) const{
+    OS << "save points:\n";
+
+    for (auto &item : SavePoints.get())
+      OS << printMBBReference(*item.first) << "\n";
```
```suggestion
      OS << printMBBReference(*item.first) << '\n';
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/MachineFrameInfo.cpp:
```diff
@@ -244,6 +244,23 @@ void MachineFrameInfo::print(const MachineFunction &MF, raw_ostream &OS) const{
+  if (!RestorePoints.empty()) {
+    OS << "restore points:\n";
+    for (auto &item : RestorePoints.get())
+      OS << printMBBReference(*item.first) << "\n";
```
```suggestion
      OS << printMBBReference(*item.first) << '\n';
```

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,57 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+static void fillCSInfoPerBB(MachineFrameInfo &MFI,
+                            DenseMap<Register, CalleeSavedInfo *> &RegToInfo,
+                            MBBVector &PrologEpilogBlocks, bool isSave) {
+  std::vector<CalleeSavedInfo> CSIV = {};
+  std::vector<CalleeSavedInfo> GCSIV = {};
```
```suggestion
  std::vector<CalleeSavedInfo> CSIV;
  std::vector<CalleeSavedInfo> GCSIV;
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,57 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+          }) != std::end(GCSIV))
+        continue;
+      for (auto BB : PrologEpilogBlocks) {
+        if (Inner.contains(BB)) {
+          Inner[BB].push_back(*RTI.second);
```
Avoid the double map lookup

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,57 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+      for (auto BB : PrologEpilogBlocks) {
+        if (Inner.contains(BB)) {
+          Inner[BB].push_back(*RTI.second);
+          std::sort(Inner[BB].begin(), Inner[BB].end(),
```
Quadruple map lookup

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -678,18 +757,57 @@ void PEI::spillCalleeSavedRegs(MachineFunction &MF) {
+      RegToInfo.insert({CS.getReg(), &CS});
+
+    if (!MFI.getSavePoints().empty()) {
+      fillCSInfoPerBB(MFI, RegToInfo, PrologBlocks, true /* isSave */);
+      fillCSInfoPerBB(MFI, RegToInfo, EpilogBlocks, false /* isSave */);
```
```suggestion
      fillCSInfoPerBB(MFI, RegToInfo, PrologBlocks, /*isSave=*/true);
      fillCSInfoPerBB(MFI, RegToInfo, EpilogBlocks, /*isSave=*/false);
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -678,18 +757,57 @@ void PEI::spillCalleeSavedRegs(MachineFunction &MF) {
+        SaveRestorePoints::PointsMap RestorePts;
+        for (MachineBasicBlock *EpilogBlock : EpilogBlocks)
+          RestorePts.insert({EpilogBlock, MFI.getCalleeSavedInfo()});
+        MFI.setRestorePoints(RestorePts);
```
std::move, or just have MFI directly manage each insertion?

enoskova-sc replied:
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,99 @@ class MachineFrameInfo {
+    SaveRestorePoints Pts{};
+    for (auto &Src : SRP) {
+      Pts.insert(std::make_pair(BBMap.find(Src.first)->second, Src.second));
+    }
```
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,99 @@ class MachineFrameInfo {
+    return Pts;
+  }
+
+  void setRestorePoint(MachineBasicBlock *MBB, std::vector<Register> &Regs) {
```
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -72,6 +89,16 @@ class CalleeSavedInfo {
+  void setSpilledIn(std::vector<MachineBasicBlock *> BBV) {
+    SpilledIn = std::move(BBV);
+  }
+  void setRestoredIn(std::vector<MachineBasicBlock *> BBV) {
```
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -37,6 +52,8 @@ class CalleeSavedInfo {
     int FrameIdx;
     unsigned DstReg;
   };
+  std::vector<MachineBasicBlock *> SpilledIn;
```
Adressed. Now we have one structure called SaveRestorePoints.

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,99 @@ class MachineFrameInfo {
+    RestorePoints = std::move(NewRestorePoints);
+  }
+
+  void setSavePoint(MachineBasicBlock *MBB, const std::vector<Register> &Regs) {
```
addressed

enoskova-sc commented on llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:
```diff
@@ -1867,13 +1910,13 @@ bool RISCVFrameLowering::spillCalleeSavedRegisters(
     return true;
 
   MachineFunction *MF = MBB.getParent();
+  auto *RVFI = MF->getInfo<RISCVMachineFunctionInfo>();
```
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -72,6 +89,16 @@ class CalleeSavedInfo {
+  ArrayRef<MachineBasicBlock *> restoredIn() const { return RestoredIn; }
+  void addSpilledIn(MachineBasicBlock *MBB) { SpilledIn.push_back(MBB); }
+  void addRestoredIn(MachineBasicBlock *MBB) { RestoredIn.push_back(MBB); }
+  void setSpilledIn(std::vector<MachineBasicBlock *> BBV) {
```
no longer actual

enoskova-sc commented:
@michaelmaitland, @arsenm, any more comments?

arsenm commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -631,6 +631,79 @@ LLVM_YAML_IS_SEQUENCE_VECTOR(llvm::yaml::CalledGlobal)
+// 1. Without explicit register enumeration:
+//      savePoint:       '%bb.n'
+//      restorePoint:    '%bb.n'
+// supported for backward compatibility (in this case we assume that all
+// CalleeSavedRegisters are splilled/restored in these points)
+// 2. With explicit register:
```
There's no expectation of MIR backwards compatibility

enoskova-sc replied:
Removed info about backward compatibility from comment.

arsenm commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -631,6 +631,79 @@ LLVM_YAML_IS_SEQUENCE_VECTOR(llvm::yaml::CalledGlobal)
 namespace yaml {
 
+// Struct representing one save/restore point in the
+// 'savePoint' / 'restorePoint' list. One point consist of machine basic block
```
```suggestion
// 'savePoint' / 'restorePoint' list. One point consists of machine basic block
```

enoskova-sc replied:
addressed

arsenm commented on llvm/include/llvm/CodeGen/MIRYamlMapping.h:
```diff
@@ -631,6 +631,79 @@ LLVM_YAML_IS_SEQUENCE_VECTOR(llvm::yaml::CalledGlobal)
+//      savePoint:       '%bb.n'
+//      restorePoint:    '%bb.n'
+// supported for backward compatibility (in this case we assume that all
+// CalleeSavedRegisters are splilled/restored in these points)
```
We ought to make callee saved registers an explicit mask operand on return instructions

arsenm commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -646,6 +647,41 @@ void MIRPrinter::convert(yaml::MachineFunction &MF,
+    StrOS << printMBBReference(*MBBEntry.first);
+    Entry.Point = StrOS.str().str();
+    Str.clear();
+    for (auto &Reg : MBBEntry.second) {
```
no auto

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -646,6 +647,41 @@ void MIRPrinter::convert(yaml::MachineFunction &MF,
+    Entry.Point = StrOS.str().str();
+    Str.clear();
+    for (auto &Reg : MBBEntry.second) {
+      if (Reg.getReg() != MCRegister::NoRegister) {
```
```suggestion
      if (Reg.getReg()) {
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,67 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+            return CSI.getReg() == RTI.first;
+          }) != std::end(GCSIV))
+        continue;
+      for (auto BB : PrologEpilogBlocks) {
```
no auto

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -478,12 +478,59 @@ bool RISCVRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
+
+      if (MI.mayStore())
+        SpilledIn = MFI.findSpilledIn(*It);
+
```
```suggestion
```

enoskova-sc replied:
addressed

arsenm commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -478,12 +478,59 @@ bool RISCVRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
+      if (MI.mayStore())
+        SpilledIn = MFI.findSpilledIn(*It);
+
+      else if (MI.mayLoad())
```
what about may load and may store? What about the else case? Should that be unreachable?

enoskova-sc replied:
I can't imagine instruction in RISC-V, that load and store simultaneously on frame index, the same for instructions, which doesn't load and doesn't store on frame index. 
Added assert and unreachable.

enoskova-sc replied:
Actually, I can't imagine situation, in which atomic instructions are used to work with frame index. 
But may be you have an example?

ppenzin commented:
@preames @arsenm any unaddressed feedback? There is a note about merging this PR and the next one in series (because it introduces registers into tests), but IMO reducing scope of the review is probably preferable.

ppenzin reviewed (approved):
This looks good to me, curious what others think.

enoskova-sc commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,67 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+        }
+        CSIV.clear();
+        CSIV.push_back(*RTI.second);
+        Inner.insert(std::make_pair(BB, CSIV));
```
addressed

enoskova-sc commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1077,6 +1075,52 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+      return false;
+    if (parseMBBReference(PFS, MBB, StringRepr))
+      return true;
+    SRPoints.insert(std::make_pair(MBB, MFI.getCalleeSavedInfo()));
```
addressed

enoskova-sc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -74,6 +74,36 @@ class CalleeSavedInfo {
+public:
+  const PointsMap &get() const { return Map; }
+
+  const std::vector<CalleeSavedInfo> getCSInfo(MachineBasicBlock *MBB) const {
```
addressed

enoskova-sc commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,67 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+        if (auto Entry = Inner.find(BB); Entry != Inner.end()) {
+          auto &CSI = Entry->second;
+          CSI.push_back(*RTI.second);
+          std::sort(CSI.begin(), CSI.end(),
```
addressed

enoskova-sc commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -651,6 +679,67 @@ static void insertCSRRestores(MachineBasicBlock &RestoreBlock,
+    }
+    // We need to sort CSIV, because Aarch64 expect CSI list to come sorted by
+    // frame index
+    std::sort(CSIV.begin(), CSIV.end(),
```
addressed

enoskova-sc commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1077,6 +1075,53 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+      return false;
+
+    for (const auto &Entry : VectorRepr) {
+      const auto &MBBSource = Entry.Point;
```
addressed

enoskova-sc commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -646,6 +647,41 @@ void MIRPrinter::convert(yaml::MachineFunction &MF,
+      }
+    }
+    // Sort here needed for stable output for lit tests
+    std::sort(Entry.Registers.begin(), Entry.Registers.end(),
```
Unfortunately, llvm::sort doesn't work for fields of structures, created in yaml namespace.
yaml namespace has its own yaml::begin() and due to it call to begin() in adl_begin() becomes ambiguous.

enoskova-sc commented:
@topperc, @arsenm, what should be done to merge this PR?

preames reviewed (commented):
A couple of suggestions on how to move this forward.  This is premised on the fact that we have general agreement on the direction, we're just splitting the code apart for ease of review.

Please split the review - in particular, remove all the target specific lowering and prolog epilogue into their own individual reviews.  This change should *only* introduce the ability to round trip the new MIR format.  Its fine to have client code *assert* there's only one save/restore point when exercised for the moment, and most of the API changes can be deferred to future patches.

In fact, the smallest possible change might be a parser/printer change which changes the MIR textual representation, but *doesn't* actually change the internal ability to represent more than one save/restore point.  All the existing tests and code paths work with one point, you can wrap that into the textual parsing for the array syntax, and it should only be a few dozen lines of code + some test updates.  

Another possible split point in the patch is splitting the notion of multiple save/restore points (with the same set of CSRs) from being able to split a different set of CSRs in each point.  This would require the notion of multiple *points* but would not require any the logic for tracking individual registers.

preames commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -332,10 +332,10 @@ class MachineFrameInfo {
-  /// Not null, if shrink-wrapping found a better place for the epilogue.
-  MachineBasicBlock *Restore = nullptr;
+  /// Not empty, if shrink-wrapping found a better place for the prologue.
+  std::vector<MachineBasicBlock *> SavePoints;
```
Why not SmallVector?  This seems like a classic use case.

preames commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -825,10 +825,21 @@ class MachineFrameInfo {
+  const std::vector<MachineBasicBlock *> &getSavePoints() const {
+    return SavePoints;
+  }
+  void setSavePoints(std::vector<MachineBasicBlock *> NewSavePoints) {
```
Use an ArrayRef instead, construct the SmallVector from it.

preames commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -825,10 +825,21 @@ class MachineFrameInfo {
+    RestorePoints = std::move(NewRestorePoints);
+  }
+
+  void clearSavePoints() { SavePoints.clear(); }
```
You don't need these, you can use setSavePoints({});

preames commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1093,6 +1091,44 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+// Return true if basic block was incorrectly specified in MIR
+bool MIRParserImpl::initializeSaveRestorePoints(
+    PerFunctionMIParsingState &PFS, const yaml::SaveRestorePoints &YamlSRPoints,
+    bool IsSavePoints) {
```
Adjust the interface to return a SmallVector by reference-param, and move the calls on MFI to the caller.

enoskova-sc replied:
addressed

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -423,16 +423,20 @@ void PEIImpl::calculateCallFrameInfo(MachineFunction &MF) {
-    assert(MFI.getRestorePoint() && "Both restore and save must be set");
-    MachineBasicBlock *RestoreBlock = MFI.getRestorePoint();
+  if (!MFI.getSavePoints().empty()) {
+    assert(MFI.getSavePoints().size() < 2 &&
```
== 1

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -423,16 +423,20 @@ void PEIImpl::calculateCallFrameInfo(MachineFunction &MF) {
-    MachineBasicBlock *RestoreBlock = MFI.getRestorePoint();
+  if (!MFI.getSavePoints().empty()) {
+    assert(MFI.getSavePoints().size() < 2 &&
+           "MFI can't contain multiple save points!");
```
Replace with "multiple save points are not yet supported"

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -423,16 +423,20 @@ void PEIImpl::calculateCallFrameInfo(MachineFunction &MF) {
+    assert(MFI.getSavePoints().size() < 2 &&
+           "MFI can't contain multiple save points!");
+    SaveBlocks.push_back(MFI.getSavePoints().front());
+    assert(!MFI.getRestorePoints().empty() &&
```
Remove

enoskova-sc replied:
addressed

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -423,16 +423,20 @@ void PEIImpl::calculateCallFrameInfo(MachineFunction &MF) {
+    SaveBlocks.push_back(MFI.getSavePoints().front());
+    assert(!MFI.getRestorePoints().empty() &&
+           "Both restore and save must be set");
+    assert(MFI.getRestorePoints().size() < 2 &&
```
== 1 (also subsumes previous assert)

enoskova-sc replied:
addressed

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -441,13 +445,15 @@ void PEIImpl::calculateSaveRestoreBlocks(MachineFunction &MF) {
-      SaveBlocks.push_back(&MBB);
-    if (MBB.isReturnBlock())
-      RestoreBlocks.push_back(&MBB);
+  if (MFI.getSavePoints().empty()) {
```
You have an early return just above.  Delete the if-clause and reduce indentation.

enoskova-sc replied:
addressed

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -558,7 +564,11 @@ static void updateLiveness(MachineFunction &MF) {
-  MachineBasicBlock *Save = MFI.getSavePoint();
+
+  assert(MFI.getSavePoints().size() < 2 &&
+         "MFI can't contain multiple save points!");
```
replace: "Multiple save points not yet supported"

preames commented on llvm/lib/CodeGen/PrologEpilogInserter.cpp:
```diff
@@ -569,7 +579,10 @@ static void updateLiveness(MachineFunction &MF) {
 
-  MachineBasicBlock *Restore = MFI.getRestorePoint();
+  assert(MFI.getRestorePoints().size() < 2 &&
+         "MFI can't contain multiple restore points!");
```
Same

enoskova-sc replied:
addressed

preames commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -209,10 +209,15 @@ void SILowerSGPRSpills::calculateSaveRestoreBlocks(MachineFunction &MF) {
-    MachineBasicBlock *RestoreBlock = MFI.getRestorePoint();
+  if (!MFI.getSavePoints().empty()) {
+    assert(MFI.getSavePoints().size() < 2 &&
+           "MFI can't contain multiple save points!");
```
Same basic pattern to comments as above...

enoskova-sc replied:
addressed

preames commented on llvm/tools/llvm-reduce/ReducerWorkItem.cpp:
```diff
@@ -92,11 +92,23 @@ static void cloneFrameInfo(
-    DstMFI.setRestorePoint(Src2DstMBB.find(RestorePt)->second);
+  assert(SrcMFI.getSavePoints().size() < 2 &&
+         "MFI can't contain multiple save points!");
+  if (!SrcMFI.getSavePoints().empty()) {
```
This code isn't correct or stylistically a good idea.  Handle the general case of an array of save/restore points, and the incorrect bit (incorrect handling of empty source) falls out for free.

enoskova-sc replied:
addressed

preames reviewed (commented):
Comments inline.  With these addressed, I plan to approve as this has been on review for a long time, and we really need to get this moving.

enoskova-sc committed:
review changes

enoskova-sc committed:
review

preames reviewed (approved):
LGTM

For context of other reviewers....

This patch is not perfect (see below), but it's a reasonable step and has been stuck in review for an extremely long time.  I am personally committing to following up on a few of the nits so that this can make forward progress.

A couple points for the record:
* This currently only models the case where *all* CSRs are spill/filled at the same set of points.  This was a deliberate simplification for this patch, and will likely be revisited in future patches.
* This currently parses both the old MIR syntax and the new array syntax.  This was triggered by an old review comment by me, and I think everyone agrees this was a bad idea in terms of net complexity.  Once this lands, I'll post a patch which removes the variant bits, and updates the existing MIR tests in one go.  

Once this goes in, we can move forward with reviewing the actual changes to shrink wrapping.

preames commented:
I've pushed a build fix in d8ce19a.  With this check-llvm passes.

enoskova-sc commented:
Fix for failures: https://github.com/llvm/llvm-project/pull/153198

preames commented:
std::variant cleanup posted: https://github.com/llvm/llvm-project/pull/153226

@enoskova-sc By some chance, are you not a member of the llvm-project github organization?  I was unable to add you as a reviewer for the change.  If not, can you fix that please?
